core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -297,10 +418,16 @@ class LogSegment(val log: FileMessageSet,
   def lastModified = log.file.lastModified
 
   /**
+   * The largest timestamp this segment contains.
+   */
+  def largestTimestamp = if (maxTimestampSoFar >= 0) maxTimestampSoFar else lastModified
```

Comment:
Yes, after the producer upgrades to 0.10, all new messages will have a timestamp and the time-based retention will be more accurate. For the old messages, the current behavior is to use the current last modified time for retention. We don't really have to improve it if that makes the new code simpler.